File: frontend/src/schemas/annotations.ts
// Shape-annotation editor schema (arrows, lines, boxes, trapezoids, ellipses drawn
// directly on the canvas). Vertices/centers are in the same data/world coordinate
// space SpatialCanvas already hands to draw interactions (PickingInfo.coordinate).
// backend/app/schemas/annotations.py is the hand-kept pydantic counterpart —
// field names and enums must match exactly.
import { z } from 'zod';

const Point = z.tuple([z.number(), z.number()]);

export const StrokeStyle = z.object({
  color: z.string(),
  width: z.number().min(0),
  dash: z.enum(['solid', 'dashed', 'dotted']),
  // "arrow" is a line whose stroke has an arrowhead at one or both ends — there is
  // no separate 'arrow' shape kind, since its geometry is identical to 'line'.
  arrowStart: z.boolean(),
  arrowEnd: z.boolean(),
  z: z.number().int(),
});
export type StrokeStyle = z.infer<typeof StrokeStyle>;

export const FillStyle = z.object({
  enabled: z.boolean(),
  color: z.string(),
  alpha: z.number().min(0).max(1),
  z: z.number().int(),
});
export type FillStyle = z.infer<typeof FillStyle>;

const LineGeometry = z.object({ kind: z.literal('line'), vertices: z.tuple([Point, Point]) });
const BoxGeometry = z.object({ kind: z.literal('box'), vertices: z.tuple([Point, Point, Point, Point]) });
const TrapezoidGeometry = z.object({ kind: z.literal('trapezoid'), vertices: z.tuple([Point, Point, Point, Point]) });
const EllipseGeometry = z.object({
  kind: z.literal('ellipse'),
  center: Point,
  radiusX: z.number().min(0),
  radiusY: z.number().min(0),
  rotation: z.number(),
});

export const ShapeGeometry = z.discriminatedUnion('kind', [
  LineGeometry,
  BoxGeometry,
  TrapezoidGeometry,
  EllipseGeometry,
]);
export type ShapeGeometry = z.infer<typeof ShapeGeometry>;
export type ShapeKind = ShapeGeometry['kind'];

export const ShapeAnnotation = z.object({
  id: z.string(),
  label: z.string().optional(),
  geometry: ShapeGeometry,
  stroke: StrokeStyle,
  // Omitted (ignored) for 'line' — a line has no interior to fill.
  fill: FillStyle.optional(),
});
export type ShapeAnnotation = z.infer<typeof ShapeAnnotation>;

export const SHAPE_KINDS: ShapeKind[] = ['line', 'box', 'trapezoid', 'ellipse'];

export function defaultStroke(): StrokeStyle {
  return { color: '#3388ff', width: 2, dash: 'solid', arrowStart: false, arrowEnd: false, z: 0 };
}

export function defaultFill(): FillStyle {
  return { enabled: true, color: '#3388ff', alpha: 0.25, z: 0 };
}

File: backend/app/schemas/annotations.py
"""Pydantic counterpart of frontend/src/schemas/annotations.ts — hand-kept in
sync field-for-field (no codegen). Validates shape-annotation job payloads
before they touch the `sdata.shapes["annotations"]` GeoDataFrame
(see sessions/shape_annotations.py).
"""
from __future__ import annotations

from typing import Literal, Union

from pydantic import BaseModel, Field
from typing_extensions import Annotated

Point = tuple[float, float]


class StrokeStyle(BaseModel):
    color: str
    width: float = Field(ge=0)
    dash: Literal["solid", "dashed", "dotted"]
    arrowStart: bool
    arrowEnd: bool
    arrowSize: float = Field(ge=0)
    z: int


class FillStyle(BaseModel):
    enabled: bool
    color: str
    alpha: float = Field(ge=0, le=1)
    z: int


class LineGeometry(BaseModel):
    kind: Literal["line"]
    vertices: tuple[Point, Point]


class BoxGeometry(BaseModel):
    kind: Literal["box"]
    vertices: tuple[Point, Point, Point, Point]


class PolygonGeometry(BaseModel):
    kind: Literal["polygon"]
    # A free-form closed ring (last vertex connects back to the first).
    vertices: list[Point] = Field(min_length=3)


class EllipseGeometry(BaseModel):
    kind: Literal["ellipse"]
    center: Point
    radiusX: float = Field(ge=0)
    radiusY: float = Field(ge=0)
    rotation: float


class TextGeometry(BaseModel):
    kind: Literal["text"]
    position: Point
    text: str
    # World-space glyph height (see the frontend TextLayer's sizeUnits: 'common'),
    # so it can be well below 1 for fine-coordinate datasets.
    fontSize: float = Field(gt=0)
    # Radians about the anchor; defaults to 0 so labels authored before rotation
    # existed still validate.
    rotation: float = 0.0


ShapeGeometry = Annotated[
    Union[LineGeometry, BoxGeometry, PolygonGeometry, EllipseGeometry, TextGeometry],
    Field(discriminator="kind"),
]


class ShapeAnnotation(BaseModel):
    id: str | None = None
    label: str | None = None
    geometry: ShapeGeometry
    stroke: StrokeStyle
    # Absent/ignored for 'line' and 'text' geometries — neither has an interior to fill.
    fill: FillStyle | None = None
